src/watch_filter.rs

```diff
@@ -0,0 +1,313 @@
+
+        WatchFilter {
+            path_patterns: path_patterns
+                .iter()
+                .map(|path| Pattern::new(app_dir.join(path).to_str().unwrap()).unwrap())
+                .collect(),
+            ignore_patterns: ignore_patterns
+                .iter()
+                .map(|path| Pattern::new(app_dir.join(path).to_str().unwrap()).unwrap())
+                .collect(),
+            process_start_since_epoch: start_time,
+        }
+    }
+
+    fn has_valid_event_kind(&self, event: &Event) -> bool {
+        event.tags.iter().any(|tag| {
+            matches!(
+                tag,
+                Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Data(_)))
+                    | Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Name(_)))
+                    | Tag::FileEventKind(FileEventKind::Create(_))
+                    | Tag::FileEventKind(FileEventKind::Remove(_))
+            )
+        })
+    }
+
+    fn matches_one_of_ignore_patterns(&self, event: &Event) -> bool {
+        event.paths().any(|(path, _)| {
+            self.ignore_patterns
+                .iter()
+                .any(|pattern| pattern.matches_path(path))
+        })
+    }
+
+    fn matches_one_of_path_patterns(&self, event: &Event) -> bool {
+        event.paths().any(|(path, _)| {
+            self.path_patterns
+                .iter()
+                .any(|pattern| pattern.matches_path(path))
+        })
+    }
+
+    /// This is necessary to check due to a bug on macOS emitting modify events on copies
+    /// https://github.com/rust-lang/rust/issues/107130
+    fn path_has_been_actually_modified(&self, event: &Event) -> Result<bool> {
+        // No need to check if a deleted path has been modified because it is gone
+        if event
+            .tags
+            .iter()
+            .any(|tag| matches!(tag, Tag::FileEventKind(FileEventKind::Remove(_))))
+        {
+            return Ok(true);
+        }
+
+        for (path, _) in event.paths() {
+            let metadata = std::fs::metadata(path)?;
+            let time_since_epoch = metadata
+                .modified()?
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_micros();
+            let mut modified_map = FILES_MODIFIED_AT.lock().unwrap();
+            let path_key = path.to_str().unwrap().to_owned();
+
+            let value = modified_map.get(&path_key);
+            match value {
+                Some(val) => {
+                    if time_since_epoch > *val {
+                        modified_map.insert(path_key, time_since_epoch);
+                        return Ok(true);
+                    }
+                }
+                None => {
+                    if time_since_epoch > self.process_start_since_epoch {
+                        modified_map.insert(path_key, time_since_epoch);
+                        return Ok(true);
+                    }
+                }
+            }
```

Comment:
A little tidier (assuming it actually works...):
```suggestion
            let base_time = modified_map
                .get(&path_key)
                .unwrap_or(&self.process_start_since_epoch);
            if time_since_epoch > *base_time {
                modified_map.insert(path_key, time_since_epoch);
                return Ok(true);
            }
```

Reply:
Noice!